sdk/servicebus/service-bus/src/core/streamingReceiver.ts

```diff
@@ -80,9 +80,14 @@ export class StreamingReceiver extends MessageReceiver {
 
   /**
    * Starts the receiver by establishing an AMQP session and an AMQP receiver link on the session.
-   **/
-  receive(): void {
+   *
+   * @param {OnMessage} onMessage The message handler to receive servicebus messages.
+   * @param {OnError} onError The error handler to receive an error that occurs while receivin messages.
+   */
+  receive(onMessage: OnMessage, onError: OnError): void {
     throwErrorIfConnectionClosed(this._context.namespace);
+    this._onMessage = onMessage;
+    this._onError = onError;
```

Comment:
The original issue was a problem occurred during setup of receiver and it was not being surfaced - I understand we fixed a specific scenario that went wrong during the setup. But later some other corner case may surface for during setup / background retries - so it seems better to have these handlers be assigned to the instance as soon as it's created and not after setup? I don't see any specific reason/benefit to doing this later? Similar comment https://github.com/Azure/azure-sdk-for-js/issues/6210#issuecomment-558274674

Reply:
> But later some other corner case may surface for during setup / background retries - so it seems better to have these handlers be assigned to the instance as soon as it's created and not after setup? I don't see any specific reason/benefit to doing this later?

Any error that happens during setup gets surfaced to the user, the receiver is not cached and no retries are made by the library

Background retries are suppose to occur only if the receiver was set up successfully and has been assigned handlers. A receiver with no handlers need not be recovered.

By assigning the handlers only in the `receive()` method, we ensure they don't get prematurely assigned.

Reply:
I agree with this. As @ramya-rao-a mentioned, errors that occur while creating the receiver get propagated to the user's error handler. Once the receiver is created and `receive` is called, the handlers still get assigned and we'll still be covered if we need to do background retries at some point.

I also like this way better because then we only keep a reference to the user's handlers if we know we can do something with them.

Reply:
Ok, got it